doc/src/manual/parallel-computing.md

```diff
@@ -1,12 +1,546 @@
+in mind, Julia natively implements interfaces to send a process through multiple cores or machines.
+Also we will mention useful external packages for distributed programming like `MPI.jl` and `DistributeArrays.jl`.
+
+
+# Coroutines
+
+Julia's parallel programming platform uses [Tasks (aka Coroutines)](@ref man-tasks) to switch among multiple
+computations. Whenever code performs a communication operation like [`fetch`](@ref) or [`wait`](@ref),
+the current task is suspended and a scheduler picks another task to run. A task is restarted when
+the event it is waiting for completes.
+
+For many problems, it is not necessary to think about tasks directly. However, they can be used
+to wait for multiple events at the same time, which provides for *dynamic scheduling*. In dynamic
+scheduling, a program decides what to compute or where to compute it based on when other jobs
+finish. This is needed for unpredictable or unbalanced workloads, where we want to assign more
+work to processes only when they finish their current tasks.
+
+As an example, consider computing the singular values of matrices of different sizes:
+
+```julia-repl
+julia> M = Matrix{Float64}[rand(800,800), rand(600,600), rand(800,800), rand(600,600)];
+
+julia> pmap(svd, M);
+```
+
+If one process handles both 800×800 matrices and another handles both 600×600 matrices, we will
+not get as much scalability as we could. The solution is to make a local task to "feed" work to
+each process when it completes its current task. For example, consider a simple [`pmap`](@ref)
+implementation:
+
+```julia
+function pmap(f, lst)
+    np = nprocs()  # determine the number of processes available
+    n = length(lst)
+    results = Vector{Any}(n)
+    i = 1
+    # function to produce the next work item from the queue.
+    # in this case it's just an index.
+    nextidx() = (global i; idx=i; i+=1; idx)
+    @sync begin
+        for p=1:np
+            if p != myid() || np == 1
+                @async begin
+                    while true
+                        idx = nextidx()
+                        idx > n && break
+                        results[idx] = remotecall_fetch(f, p, lst[idx])
+                    end
+                end
+            end
+        end
+    end
+    results
+end
+```
+
+[`@async`](@ref) is similar to [`@spawn`](@ref), but only runs tasks on the local process. We
+use it to create a "feeder" task for each process. Each task picks the next index that needs to
+be computed, then waits for its process to finish, then repeats until we run out of indices. Note
+that the feeder tasks do not begin to execute until the main task reaches the end of the [`@sync`](@ref)
+block, at which point it surrenders control and waits for all the local tasks to complete before
+returning from the function. The feeder tasks are able to share state via `nextidx` because
+they all run on the same process. No locking is required, since the threads are scheduled cooperatively
```

Comment:
I would prefer avoiding stating such an assumption in the manual. While this is true right now, it is also an implementation detail and as far as I know the plan is still to move tasks more to a `m:n` model (running `m` tasks on `n` threads) and under such a model the update to `i` would need to be locked.